### PR TITLE
MAINT: Remove `__setslice__` and `__getslice__`

### PR DESCRIPTION
**Pull request by eric-wieser**
This code was only here for Python 2.5 compatibility, but numpy requires 2.7 at minimum.

This involves:

* Removing all the `__[gs]etslice__` method from python files
* Removing the `sq_slice` and `sq_slice_ass` slots from C files, so that the python files do not have to override the base implementations

### CODE REVIEW

**Comment by eric-wieser** [comment]
This fails in a bunch of bizarre and unexpected ways?

**Review by shoyer — APPROVED** [review]

File: doc/source/reference/maskedarray.baseclass.rst [hunk 1]
```diff
@@ -417,8 +417,6 @@ Container customization: (see :ref:`Indexing <arrays.indexing>`)
    MaskedArray.__getitem__
    MaskedArray.__setitem__
    MaskedArray.__delitem__
-   MaskedArray.__getslice__
-   MaskedArray.__setslice__
```
Comment (charris) [inline doc/source/reference/maskedarray.baseclass.rst]: Note that `getslice` and `setslice` are also in the index. Grep for the terms.

**Comment by charris** [comment]
I'm a bit concerned about backward compatibility, note the reference from @mhvk. Both Pandas and SciPy look good, however, so maybe not worth worrying about.

**Comment by charris** [comment]
Needs a mention in the 1.13.0 release notes in the compatibility section.

File: doc/source/reference/maskedarray.baseclass.rst (hunk 1 above)
Comment (eric-wieser) [inline doc/source/reference/maskedarray.baseclass.rst]: Right, but without the `__`. I was under the impression that was just a coincidental reference to something else. If that's not the case, I'll wipe that one too

**Comment by eric-wieser** [comment]
I'm not convinced this would have any compatibility issues beyond `super().__setslice__` failing

**Comment by eric-wieser** [comment]
I'll do some experiments to characterize exactly what the effect is

**Comment by eric-wieser** [comment]
@charris: Changelog added. Compatibility definitely only affects Python 2.7. Experiments so far:

* Forgetting to implement `MyClass.__getslice__` on ndarray subclasses would cause `ndarray.__getslice__` to be called
  * Now: `MyClass.__getitem__` is called instead, which is already the py3 behaviour

* The above only applies because `ndarray` is a C-api class. If it were a pure-python class, that wouldn't be the case.

File: doc/source/reference/arrays.indexing.rst [hunk 2]
```diff
@@ -36,8 +36,8 @@ objects, the :const:`Ellipsis` object, or the :const:`newaxis` object,
 but not for integer arrays or other embedded sequences.
 
 .. index::
-   triple: ndarray; special methods; getslice
-   triple: ndarray; special methods; setslice
+   triple: ndarray; special methods; get slice
```
Comment (charris) [inline doc/source/reference/arrays.indexing.rst]: Should be an ndarray method. `getitem` and `setitem` are probably what should be here.

Comment (eric-wieser) [inline doc/source/reference/arrays.indexing.rst]: Why not `__getitem__` and `__setitem__
`?

**Comment by charris** [comment]
Made a small fix and will commit when tests finish. Yeah, I forgot `[ci skip]` :(

**Comment by eric-wieser** [comment]
@charris: didn't know you could edit the PR! Is that new?

**Comment by charris** [comment]
No, not new but not often used. You can disable it when you make the PR, there is a button somewhere...

**Comment by charris** [comment]
Thanks @eric-wieser .

**Comment by pv** [comment]
Performance regression?
https://pv.github.io/numpy-bench/index.html#bench_indexing.IndexingSeparate.time_mmap_slicing?commits=1b877254af0850d025cdc5d07b3fcaa1614dbe4b

**Comment by charris** [comment]
@pv Looks like that. Does that show up in Python3? If not, I think it indicates that getitem needs some improvement.

**Comment by seberg** [comment]
Frankly I doubt there is much you can do (easily). Slicing is a pretty fast operation, so simply creating a few extra python objects and more complex getitem logic may already add a substantial slowdown. But maybe its worth a look....

**Comment by eric-wieser** [comment]
Perhaps there's an argument for `__getitem__` trying to speed up the case when the input is a single slice object

File: numpy/core/src/multiarray/sequence.c [hunk 3]
```diff
@@ -27,90 +27,6 @@ array_any_nonzero(PyArrayObject *mp);
...
-                             (PyObject *)self);
-    if (ret == NULL) {
-        return NULL;
-    }
-    Py_INCREF(self);
-    if (PyArray_SetBaseObject(ret, (PyObject *)self) < 0) {
-        Py_DECREF(ret);
-        return NULL;
-    }
-    PyArray_UpdateFlags(ret, NPY_ARRAY_UPDATE_ALL);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/sequence.c]: Why does this line exist here but not in `get_view_from_index`?

**Comment by eric-wieser** [comment]
Another bonus of killing this:

```python
#python 2.7
>>> x = np.array(0)
>>> x[-1:]
TypeError: len() of unsized object
>>> x[np.s_[-1:]]
IndexError: too many indices for array
```

File: numpy/core/src/multiarray/sequence.c (hunk 3 above)
Comment (seberg) [inline numpy/core/src/multiarray/sequence.c]: Hmmm, no idea why this should be needed here. Maybe it was just being very careful, or a git blame might help to see when the line was added if it was. The array creation should set all the flags correctly already, except maybe writable, but that is correct.

**Comment by juliantaylor** [comment]
this change needs to be reverted.
There are programs out there calling setslice. E.g. astropy is massively broken.

we do not break backward compatibility for cosmetic reasons!

**Pull request merged**